cscs-checks/apps/vasp/vasp_check.py

```diff
@@ -41,41 +48,52 @@ def time(self):
 
 @rfm.simple_test
 class VASPCpuCheck(VASPCheck):
```

Comment:
Since you are changing the tests, I would rather make the `cpu` and `gpu` versions of the test a parameter.

Reply:
I have changed the checks to be selected based on the architecture: initially I used `constraint = parameter(['gpu', 'mc'])`, but it seems redundant. However now both `gpu` and `mc` are executed and it seems that I cannot exclude the first with the command line option `--cpu-only`.

Reply:
You need to set `num_gpus_per_node` in order for the `--cpu-only` and `--gpu-only` options to take effect; if I'm not mistaken.

Reply:
I have added `num_gpus_per_node` but it does not seem to be taken into account by `--cpu-only` or `--gpu-only`: anyway one can always use `--system daint:gpu` or `--system daint:mc` to select the architecture directly.

Reply:
This is strange because I just checked the filter and it indeed checks whether the `num_gpus_per_node` is greater than zero:

https://github.com/eth-cscs/reframe/blob/7b27cbe664fa50bea964202db8b0cfb61f278288/reframe/frontend/filters.py#L70-L81

And here:

https://github.com/eth-cscs/reframe/blob/7b27cbe664fa50bea964202db8b0cfb61f278288/reframe/frontend/cli.py#L954-L957

Reply:
Yes, the problem is the following. Test filtering from the command line happens after the test is initialized and *before* the test starts executing. Here you set the `num_gpus_per_node` in a pre-run hook. This is too late for the `--cpu-only` and `--gpu-only` options to take it into account. By the time of the filtering, `num_gpus_per_node` is zero, therefore they seem to have no effect. You have to set `num_gpus_per_node` in a post-init hook to be able to filter correctly.

Reply:
Thanks for your note: I set `num_gpus_per_node` in `@run_before('run')` based on `gromacs/gromacs_check.py`, where it has no effect either. I don't know how to set it in `@run_after('init')`, since `current_partition` is not yet defined there.

Reply:
You are right about the Gromacs check and we shouldn't have been setting it there. It's easily fixable in the Gromacs check because we can differentiate the type of the test using the `nb_impl` parameter and set the `num_gpus_per_node` in a post-init hook. But in the VASP test here, I see that you run both versions (GPU and CPU) identically? We could do the same trick as with Gromacs for VASP and add a parameter to differentiate the test. I think that's a better solution than relying on the software stack to do the differentiation; the test has quite a much of implicit information if running on Daint/Dom implies running the GPU version. What do you think?

Reply:
Yes, the two  checks of VASP are using the same input files. I can define `constraint = parameter(['gpu', 'mc'])`, although it is redundant, since one can choose it anyway with the command line option `--system daint:gpu` or `--system daint:mc`.

Reply:
That's the downside in this case, that the test is not different for each variant. We can keep it as is, if we are fine with the `num_gpus_per_node` filtering limitation.

Reply:
I think it is fine to leave it as it is, since we have still the option to select the `gpu` or `mc` checks if necessary. I would remove `num_gpus_per_node` since it is not having any effect in the pre-run hook, though